jsapi/src/statemanager.js

```diff
@@ -0,0 +1,94 @@
+goog.provide('lux.StateManager');
+
+
+/**
+ * @classdesc
+ * @constructor
+ */
+lux.StateManager = function() {
+
+  /**
+   * @type {lux.Map}
+   */
+  this.map_ = null;
+
+  /**
+   * @type {luxx.State}
+   */
+  this.state_ = {};
+};
+
+/**
+ * @param {lux.Map} map The map.
+ */
+lux.StateManager.prototype.setMap = function(map) {
+  this.map_ = map;
+  var view = map.getView();
+
+  function onViewUpdate() {
+    var center = view.getCenter();
+    var zoom = view.getZoom();
+    var object = {};
+    object['X'] = Math.round(center[0]);
+    object['Y'] = Math.round(center[1]);
+    object['zoom'] = zoom;
+    this.updateState(object);
+  }
+
+  // make sure the state is updated the first time
+  onViewUpdate.call(this);
+
+  ol.events.listen(view, 'propertychange',
+      lux.debounce(onViewUpdate, 250).bind(this));
+
+  var layersListenerKeys = [];
+
+  function onLayersUpdate() {
+    // remove any listener on visibility change
+    layersListenerKeys.forEach(function(key) {
+      ol.events.unlistenByKey(key);
+    });
+
+    var object = {};
+    var layers = [];
+    var opacities = [];
+    map.getLayers().forEach(function(layer, index) {
+      // first layer is bg layer
+      if (index === 0) {
+        object['bgLayer'] = layer.get('name');
+      } else {
+        layers.push(layer.get('id'));
+        opacities.push(layer.get('visible') ? layer.getOpacity() : 0);
+        layersListenerKeys[layer.get('id')] = ol.events.listen(
+          layer,
+          'change:visible',
+          onLayersUpdate,
+          this);
+      }
+    }.bind(this));
+    object['layers'] = layers.join('-');
+    object['opacities'] = opacities.join('-');
+    this.updateState(object);
+  }
+
+  ol.events.listen(map.getLayers(), ol.CollectionEventType.ADD,
+    onLayersUpdate, this);
```

Comment:
I didn't expect layer to be removed. However, you're right, this should be handled.